flink-tests/src/test/java/org/apache/flink/test/javaApiOperators/CustomDistributionITCase.java

```diff
@@ -0,0 +1,184 @@
+			@Override
+			public void mapPartition(Iterable<Tuple3<Integer, Long, String>> values, Collector<Boolean> out) throws Exception {
+				int partitionIndex = getRuntimeContext().getIndexOfThisSubtask();
+
+				for (Tuple3<Integer, Long, String> s : values) {
+					if ((s.f0 - 1) / 7 != partitionIndex) {
+						fail("Record was not correctly partitioned: " + s.toString());
+					}
+				}
+			}
+		});
+
+		result.output(new DiscardingOutputFormat()); 
+		env.execute();
+	}
+
+	@Test
+	public void testRangeWithDistribution2() throws Exception{
+		/*
+		 * Test the record partitioned rightly with two fields according to the customized data distribution
+		 */
+
+		ExecutionEnvironment env = ExecutionEnvironment.createLocalEnvironment();
+
+		DataSet<Tuple3<Integer, Long, String>> input1 = CollectionDataSets.get3TupleDataSet(env);
+		final TestDataDist dist = new TestDataDist(2);
+
+		env.setParallelism(dist.getParallelism());
+
+		DataSet<Boolean> result = DataSetUtils.partitionByRange(input1.map(new MapFunction<Tuple3<Integer, Long, String>, Tuple3<Integer, Integer, String>>() {
+			@Override
+			public Tuple3<Integer, Integer, String> map(Tuple3<Integer, Long, String> value) throws Exception {
+				return new Tuple3<>(value.f0, value.f1.intValue(), value.f2);
+			}
+		}), dist, 0, 1).mapPartition(new RichMapPartitionFunction<Tuple3<Integer, Integer, String>, Boolean>() {
+			@Override
+			public void mapPartition(Iterable<Tuple3<Integer, Integer, String>> values, Collector<Boolean> out) throws Exception {
+				int partitionIndex = getRuntimeContext().getIndexOfThisSubtask();
+
+				for (Tuple3<Integer, Integer, String> s : values) {
+					if (s.f0 <= partitionIndex * (partitionIndex + 1) / 2 ||
+							s.f0 > (partitionIndex + 1) * (partitionIndex + 2) / 2 ||
+							s.f1 - 1 != partitionIndex) {
+						fail("Record was not correctly partitioned: " + s.toString());
+					}
+				}
+			}
+		});
+
+		result.output(new DiscardingOutputFormat());
+		env.execute();
+	}
+
+	/**
+	 * The class is used to do the tests of range partition with customed data distribution.
+	 */
+	public static class TestDataDist implements DataDistribution {
+
+		private int dim;
+
+		public TestDataDist() {}
+
+		/**
+		 * Constructor of the customized distribution for range partition.
+		 * @param dim the number of the fields.
+		 */
+		public TestDataDist(int dim) {
+			this.dim = dim;
+		}
+
+		public int getParallelism() {
+			if (dim == 1) {
+				return 3;
+			}
+			return 6;
+		}
+
+		@Override
+		public Object[] getBucketBoundary(int bucketNum, int totalNumBuckets) {
+			if (dim == 1) {
+				/*
+				for the first test, the boundary is just like : 
+				(0, 7]
+				(7, 14]
+				(14, 21]
+				 */
+
+				return new Integer[]{(bucketNum + 1) * 7};
+			}
+			/*
+			for the second test, the boundary is just like : 
+			(0, 1], (0, 1]
```

Comment:
This is not a common distribution for composite range partition keys. It should look similar to this one:
```
( (1, 2), (1, 7) ]
( (1, 7), (3, 2) ]
( (3, 2), (4, 8) ]
...
```
This would result in 4 partitions:

- Partition 1: `(0,2), (1,1), (0,10), ...`
- Partition 2: `(1,3), (2,20), (1,8), ...`
- Partition 3: `(4,10), (3,5), ...`
- Partition 4: `(5,3), (8,20), (4,12), ...`